### PR TITLE
[WEBSITE-616] Implementation for custom error handling

### PR DESCRIPTION
**Pull request by aebirim**
* addition of code in request helper to handle specific 204 error
* change to controllers: constituencies, people and parties and actions: members letters, letters and current_members_letters
* change to respective tests and vcr cassettes for constituencies, people and parties

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -23,6 +23,7 @@ group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug'
 
+
```
Comment (mattrayner) [inline Gemfile:25]: Please remove this extra line :)

File: app/controllers/concerns/request_helper.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+module RequestHelper
+  # @param [Parliament::Request] request the request we are going to attempt to make
```
Comment (mattrayner) [inline app/controllers/concerns/request_helper.rb]: Please add a descriptive comment on what the method does

File: app/controllers/concerns/request_helper.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+module RequestHelper
```
Comment (mattrayner) [inline app/controllers/concerns/request_helper.rb:1]: Please add a module comment for Yard

File: app/controllers/concerns/request_helper.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  # @param [Parliament::Request] request the request we are going to attempt to make
+  # @param [Block] block block to be executed if we get a 204 error
+  #
+  # @return [Hash] a response object with the format { success: Boolean, response: Parliament::Response }
+  def self.handler(request, &block)
+    return_object = { success: false, response: nil }
+    begin
```
Comment (mattrayner) [inline app/controllers/concerns/request_helper.rb]: Please add some whitespace above here

File: app/controllers/concerns/request_helper.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+      response = request.get
+      return_object = { success: true, response: response }
+    rescue Parliament::NoContentError => e
+      Rails.logger.warn "Received 204 status from #{e}"
+      block&.call
+    end
+    return return_object
```
Comment (mattrayner) [inline app/controllers/concerns/request_helper.rb]: please add some whitespace above here and remove the `return` keyword

File: app/controllers/constituencies_controller.rb [hunk 6]
```diff
@@ -86,20 +86,27 @@ def letters
...
     letter_data = Parliament::Request.new.constituencies.a_z_letters.get
-
-    @constituencies = Parliament::Request.new.constituencies(letter).get.sort_by(:name)
     @letters = letter_data.map(&:value)
+
+    request = Parliament::Request.new.constituencies(letter)
+    response = RequestHelper.handler(request) { @constituencies = [] }
+
+    @constituencies = response[:response].sort_by(:name) if response[:success]
+
```
[the remark this answers is not in the document]
Comment (mattrayner) [reply app/controllers/constituencies_controller.rb]: Please remove this extra line

File: app/controllers/constituencies_controller.rb [hunk 7]
```diff
@@ -86,20 +86,27 @@ def letters
...
-    @constituencies = data.filter('http://id.ukpds.org/schema/ConstituencyGroup')
-    @constituencies = @constituencies.sort_by(:name)
     @letters = letter_data.map(&:value)
+
+    request = Parliament::Request.new.constituencies.current(letter)
+    response = RequestHelper.handler(request) { @constituencies = [] }
+
+    @constituencies = response[:response].filter('http://id.ukpds.org/schema/ConstituencyGroup')
+    @constituencies = @constituencies.sort_by(:name) if response[:success]
+
```
[the remark this answers is not in the document]
Comment (mattrayner) [reply app/controllers/constituencies_controller.rb]: Please follow Hound

File: app/controllers/parties_controller.rb [hunk 8]
```diff
@@ -54,37 +54,43 @@ def current_members
...
     letter_data = Parliament::Request.new.parties.a_z_letters.get
-
-    @parties = data.filter('http://id.ukpds.org/schema/Party').sort_by(:name)
     @letters = letter_data.map(&:value)
+
+    request = Parliament::Request.new.parties(letter)
+    response = RequestHelper.handler(request) { @people = [] }
+
+    @parties = response[:response].filter('http://id.ukpds.org/schema/Party').sort_by(:name) if response[:success]
+
```
[the remark this answers is not in the document]
Comment (mattrayner) [reply app/controllers/parties_controller.rb]: Please follow Hound

File: app/controllers/parties_controller.rb [hunk 9]
```diff
@@ -54,37 +54,43 @@ def current_members
...
-    data = Parliament::Request.new.parties(party_id).members(letter).get
     letter_data = Parliament::Request.new.parties(party_id).members.a_z_letters.get
+    @letters = letter_data.map(&:value)
 
-    @party, @people = data.filter('http://id.ukpds.org/schema/Party', 'http://id.ukpds.org/schema/Person')
+    request = Parliament::Request.new.parties(party_id).members(letter)
+    response = RequestHelper.handler(request) { @people = [] }
+    @party, @people = response[:response].filter('http://id.ukpds.org/schema/Party', 'http://id.ukpds.org/schema/Person') if response[:success]
     @party = @party.first
     @people = @people.sort_by(:sort_name)
```
Comment (mattrayner) [inline app/controllers/parties_controller.rb]: These two lines need to be wrapped in the same conditional as: 

`@party, @people = response[:response].filter('http://id.ukpds.org/schema/Party', 'http://id.ukpds.org/schema/Person') if response[:success]`

If not, this page could still generate a 500 error

File: app/controllers/parties_controller.rb [hunk 10]
```diff
@@ -54,37 +54,43 @@ def current_members
...
     letter = params[:letter]
     party_id = params[:party_id]
 
-    data = Parliament::Request.new.parties(party_id).members(letter).get
     letter_data = Parliament::Request.new.parties(party_id).members.a_z_letters.get
+    @letters = letter_data.map(&:value)
 
-    @party, @people = data.filter('http://id.ukpds.org/schema/Party', 'http://id.ukpds.org/schema/Person')
+    request = Parliament::Request.new.parties(party_id).members(letter)
+    response = RequestHelper.handler(request) { @people = [] }
```
Comment (mattrayner) [inline app/controllers/parties_controller.rb]: Should this proc also include `@party = []`?

I'm unsure of how/if this page will display without @party?

Comment (mattrayner) [reply app/controllers/parties_controller.rb]: You will need to test this further

File: app/controllers/constituencies_controller.rb [hunk 11]
```diff
@@ -86,20 +86,27 @@ def letters
...
-    data = Parliament::Request.new.constituencies.current(letter).get
-
-    @constituencies = data.filter('http://id.ukpds.org/schema/ConstituencyGroup')
-    @constituencies = @constituencies.sort_by(:name)
     @letters = letter_data.map(&:value)
+
+    request = Parliament::Request.new.constituencies.current(letter)
+    response = RequestHelper.handler(request) { @constituencies = [] }
+
+    @constituencies = response[:response].filter('http://id.ukpds.org/schema/ConstituencyGroup')
```
Comment (mattrayner) [inline app/controllers/constituencies_controller.rb]: these two lines could be confined into:

```
@constituencies = response[:response].filter('http://id.ukpds.org/schema/ConstituencyGroup').sort_by(:name) if response[:success]
```

File: app/controllers/constituencies_controller.rb [hunk 12]
```diff
@@ -86,20 +86,27 @@ def letters
...
+    @constituencies = @constituencies.sort_by(:name) if response[:success]
```
Comment (mattrayner) [inline app/controllers/constituencies_controller.rb]: with the above change this line is not needed

File: app/controllers/parties_controller.rb [hunk 13]
```diff
@@ -54,37 +54,43 @@ def current_members
...
     letter_data = Parliament::Request.new.parties(party_id).members.current.a_z_letters.get
 
-    @party, @people = data.filter('http://id.ukpds.org/schema/Party', 'http://id.ukpds.org/schema/Person')
+    @letters = letter_data.map(&:value)
+    request = Parliament::Request.new.parties(party_id).members.current(letter)
+    response = RequestHelper.handler(request) { @people = [] }
+
+    @party, @people = response[:response].filter('http://id.ukpds.org/schema/Party', 'http://id.ukpds.org/schema/Person') if response[:success]
     @party = @party.first
     @people = @people.sort_by(:sort_name)
```
Comment (mattrayner) [inline app/controllers/parties_controller.rb]: Same as above - how does this work without @party?

File: app/controllers/people_controller.rb [hunk 14]
```diff
@@ -130,31 +130,39 @@ def current_house
...
     letter_data = Parliament::Request.new.people.a_z_letters.get
 
-    @people = Parliament::Request.new.people(letter).get.sort_by(:sort_name)
     @letters = letter_data.map(&:value)
+
+    request = Parliament::Request.new.people(letter)
+    response = RequestHelper.handler(request) { @people = [] }
+
+    @people = response[:response].sort_by(:sort_name) if response[:success]
+
```
[the remark this answers is not in the document]
Comment (mattrayner) [reply app/controllers/people_controller.rb]: please follow

File: app/controllers/people_controller.rb [hunk 15]
```diff
@@ -130,31 +130,39 @@ def current_house
...
   end
 
   def members_letters
     letter = params[:letter]
-
-    data = Parliament::Request.new.people.members(letter).get
     letter_data = Parliament::Request.new.people.members.a_z_letters.get
```
Comment (mattrayner) [inline app/controllers/people_controller.rb]: remove this line?

File: app/controllers/people_controller.rb [hunk 16]
```diff
@@ -130,31 +130,39 @@ def current_house
...
+    response = RequestHelper.handler(request) { @people = [] }
+
+    @people = response[:response].filter('http://id.ukpds.org/schema/Person').sort_by(:sort_name) if response[:success]
   end
 
   def current_members_letters
     letter = params[:letter]
-
-    data = Parliament::Request.new.people.members.current(letter).get
     letter_data = Parliament::Request.new.people.members.current.a_z_letters.get
```
Comment (mattrayner) [inline app/controllers/people_controller.rb]: Remove this line?

File: app/views/houses/members_letters.html.haml [hunk 17]
```diff
@@ -6,8 +6,11 @@
 %h1= page_title
 
 = render partial: 'shared/id_a_to_z', locals: { route_symbol: :house_members_a_z_letter_path, id: @house.graph_id}
-%ul.list
-  = render partial: 'pugin/cards/person-list', locals: { people: @people }
+- if @people.empty?
```
Comment (mattrayner) [inline app/views/houses/members_letters.html.haml]: Reset this file back to how it was before - the front-end team have already implemented this inside of 'pugin/cards/person-list'

File: app/views/shared/204.html.haml [hunk 18]
```diff
@@ -0,0 +1 @@
+%h1= "This shows no content"
```
Comment (mattrayner) [inline app/views/shared/204.html.haml]: This file shouldn't be needed any more?

File: config/environments/development.rb [hunk 19]
```diff
@@ -13,7 +13,7 @@
   config.eager_load = false
 
   # Show full error reports and disable caching.
-  config.consider_all_requests_local       = true
+  config.consider_all_requests_local       = false
```
Comment (mattrayner) [inline config/environments/development.rb]: This should be back to true with our fix

**Review by mattrayner — CHANGES_REQUESTED** [review]
Please see the blow changes

File: app/controllers/concerns/request_helper.rb [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+  # @param [Parliament::Request] request the request we are going to attempt to make
+  # @param [Block] block block to be executed if we get a 204 error
+  #
+  # @return [Hash] a response object with the format { success: Boolean, response: Parliament::Response }
+  def self.handler(request, &block)
+    return_object = { success: false, response: nil }
+    begin
+      response = request.get
+      return_object = { success: true, response: response }
+    rescue Parliament::NoContentError => e
```
Comment (mattrayner) [inline app/controllers/concerns/request_helper.rb]: With the new update, this is now `Parliament::NoContentResponseError`

Comment (aebirim) [reply app/controllers/concerns/request_helper.rb]: now updated

File: config/environments/development.rb (hunk 19 above)
Comment (aebirim) [reply config/environments/development.rb]: now updated

File: app/views/shared/204.html.haml (hunk 18 above)
Comment (aebirim) [reply app/views/shared/204.html.haml]: file removed

File: app/controllers/people_controller.rb (hunk 16 above)
Comment (aebirim) [reply app/controllers/people_controller.rb]: removed

File: app/controllers/people_controller.rb (hunk 15 above)
Comment (aebirim) [reply app/controllers/people_controller.rb]: removed

File: app/controllers/parties_controller.rb (hunk 9 above)
Comment (aebirim) [reply app/controllers/parties_controller.rb]: wrapped in conditional

File: app/controllers/parties_controller.rb (hunk 13 above)
Comment (aebirim) [reply app/controllers/parties_controller.rb]: @party now added

File: app/controllers/parties_controller.rb (hunk 10 above)
Comment (aebirim) [reply app/controllers/parties_controller.rb]: completed

File: app/controllers/constituencies_controller.rb (hunk 11 above)
Comment (aebirim) [reply app/controllers/constituencies_controller.rb]: completed

File: app/controllers/concerns/request_helper.rb (hunk 5 above)
Comment (aebirim) [reply app/controllers/concerns/request_helper.rb]: completed

File: app/controllers/concerns/request_helper.rb (hunk 4 above)
Comment (aebirim) [reply app/controllers/concerns/request_helper.rb]: completed

File: app/controllers/concerns/request_helper.rb (hunk 3 above)
Comment (aebirim) [reply app/controllers/concerns/request_helper.rb:1]: completed

File: app/controllers/concerns/request_helper.rb (hunk 2 above)
Comment (aebirim) [reply app/controllers/concerns/request_helper.rb]: completed

File: app/controllers/constituencies_controller.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (aebirim) [reply app/controllers/constituencies_controller.rb]: completed

File: app/controllers/people_controller.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (aebirim) [reply app/controllers/people_controller.rb]: completed

**Comment by mattrayner** [comment]
No longer required - covered in #338

**Pull request closed without merging**